File: java/src/test/java/testingil/webinar/tddapi/s19/refactor_navigator/NavigatorAPITests.java
package testingil.webinar.tddapi.s19.refactor_navigator;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.lessThan;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import testingil.webinar.tddapi.s19.Distance;
import testingil.webinar.tddapi.s19.DistanceProvider;
import testingil.webinar.tddapi.s19.Location;
import testingil.webinar.tddapi.s19.NavigationController19;

@WebMvcTest
@ContextConfiguration(classes = NavigatorAPITestsConfig19.class)
class NavigatorAPITests {

	MockMvc mockMvc;
	@Autowired NavigationController19 navController;
	@Autowired DistanceProvider mockProvider;

	@BeforeEach
	public void setup() {
		this.mockMvc = MockMvcBuilders
				.standaloneSetup(navController).build();
	}

	@Test
	void navigate_to_same_location_distance_is_zero() throws Exception {
		Distance distance = new Distance();
		distance.value = 0;
		
		when(mockProvider.getDistance(
				Mockito.any(Location.class), 
				Mockito.any(Location.class)))
			.thenReturn(distance);

		
		Location location = new Location("New York");
		setStartPoint(location);
		setDestination(location);
		assertThat(distanceToDestination(), is(0));
	}

	@Test
	void navigate_and_drive_to_another_location_distance_is_reduced() throws Exception {
		Distance initDistance = new Distance();
		initDistance.value = 2500;
		
		Distance midDistance = new Distance();
		midDistance.value = 1400;
		
		when(mockProvider.getDistance(
				Mockito.any(Location.class), 
				Mockito.any(Location.class)))
			.thenReturn(initDistance)
			.thenReturn(midDistance);

		Location initialLocation = new Location("New York");
		setStartPoint(initialLocation);
		Location destination = new Location("Los Angeles");;
		setDestination(destination);
		int initialDistance = distanceToDestination();
		Location midLocation = new Location("Dallas");
		driveTo(midLocation); 
		assertThat(distanceToDestination(), is(lessThan(initialDistance)));
	}

	
	
	private void setDestination(Location location) throws JsonProcessingException, Exception {
		mockMvc.perform(post("/navigator19/destination")
				.contentType(MediaType.APPLICATION_JSON)
				.content(toJson(location)))
        .andExpect(status().isOk());
	}

	private void setStartPoint(Location location) throws Exception {
		mockMvc.perform(post("/navigator19/startpoint")
				.contentType(MediaType.APPLICATION_JSON)
				.content(toJson(location)))
        .andExpect(status().isOk());
	}

	private void driveTo(Location midLocation) throws Exception {
		setStartPoint(midLocation);
	}
	
	private int distanceToDestination() throws JsonProcessingException, Exception {
		MvcResult result = 
			mockMvc.perform(get("/navigator19/distance"))
	        .andExpect(status().isOk())
	        .andReturn();
		
		String content = result.getResponse().getContentAsString();
		Distance distance = distanceFromJson(content);
		return distance.inKm();
	}

	private Distance distanceFromJson(String content) throws Exception {
		ObjectMapper objectMapper = new ObjectMapper();
		return objectMapper.readValue(content, Distance.class);
	}

	private String toJson(Location loc) throws Exception {
		ObjectMapper objectMapper = new ObjectMapper();
		return objectMapper.writeValueAsString(loc);
	}

}

File: Python/s16_pass_unit_test/navigator_api_tests.py
import json
import unittest

from hamcrest import assert_that, equal_to, less_than

from s16_pass_unit_test.distance import Distance
from s16_pass_unit_test.navigation_controller import app
from s16_pass_unit_test.location import Location


class NavigatorAPITests(unittest.TestCase):
    def test_navigate_to_same_location_distance_is_zero(self):
        location = Location("New York")
        self.set_start_point(location)
        self.setDestination(location)
        assert_that(self.distanceToDestination(), equal_to(0))

    def test_navigate_and_drive_to_another_location_distance_is_reduced(self):
        initialLocation = Location("New York")
        self.set_start_point(initialLocation)

        destination = Location("Los Angeles")
        self.setDestination(destination)

        initialDistance = self.distanceToDestination()

        midLocation = Location("Dallas")
        self.driveTo(midLocation)

        assert_that(self.distanceToDestination(), equal_to(less_than(initialDistance)))

    def set_start_point(self, location):
        response = self.app.post("/nav/startpoint",
                                 content_type='application/json',
                                 data=json.dumps(location.__dict__))
        self.assertEqual(200, response.status_code)

    def setDestination(self, location):
        response = self.app.post("/nav/destination",
                                 content_type='application/json',
                                 data=json.dumps(location.__dict__))
        self.assertEqual(200, response.status_code)

    def distanceToDestination(self):
        response = self.app.get("/nav/distance")
        self.assertEqual(200, response.status_code)
        distance = Distance(response.get_json()['miles'])
        return distance.inKm()

    def driveTo(self, location):
        self.set_start_point(location)

    def setUp(self):
        self.app = app.test_client(self)


if __name__ == '__main__':
    unittest.main()
